arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt

```diff
@@ -1,752 +1,1066 @@
+    this@list.mapRose { Rose(it) }.replicate(n)
+  }.mapRose { r ->
+    r.flatMap {
+      it.asSequence().interleave()
     }
+  }
+    .map { it.toList() }
+    .ensure { it.size >= range.lowerBound(s) }
+}
 
-    fun <K, A> Kind<M, Tuple2<K, A>>.map(range: Range<Int>): Kind<M, Map<K, A>> = sized { s ->
-        MM().run {
-            fx.monad {
-                val k = !int_(range)
-                !this@map.uniqueByKey(k)
-            }.shrink { it.shrink() }
-                .flatMap { it.sequence(MM()).map { it.fix() } }
-                .map { it.toMap() }
-                .ensure { it.size >= range.lowerBound(s) }
+/**
+ * Generate a [List] of [A]'s but uses an [IntRange] for convenience.
+ */
+fun <R, A> Gen<R, A>.list(range: IntRange): Gen<R, List<A>> =
+  list(Range.constant(range))
+
+internal fun <R, A> Gen<R, A>.replicate(n: Int): Gen<R, List<A>> =
+  if (n <= 0) Gen.just(emptyList())
+  else (0 until n).toList().fold(Gen.just(emptyList<A>()) as Gen<R, List<A>>) { acc, _ ->
+    acc.map2(this@replicate) { a, b -> a + b }
+  }
+
+internal fun <A> Sequence<A>.splits(): Sequence<Tuple3<Sequence<A>, A, Sequence<A>>> =
+  firstOrNull()?.let { x ->
+    sequenceOf(Tuple3(emptySequence<A>(), x, drop(1)))
+      // flatMap for added laziness
+      .flatMap {
+        sequenceOf(it) + drop(1).splits().map { (a, b, c) ->
+          Tuple3(sequenceOf(x) + a, b, c)
         }
-    }
+      }
+  } ?: emptySequence()
+
+internal fun <A> Sequence<Rose<A>>.dropSome(): Sequence<Rose<Sequence<A>>> =
+  toList().let { xs ->
+    if (xs.isEmpty()) emptySequence()
+    else iterate(xs.size) { it / 2 }.takeWhile { it > 0 }.flatMap { n -> xs.removes(n) }
+  }.map { it.asSequence().interleave() }
+
+internal fun <A> Sequence<Rose<A>>.shrinkOne(): Flow<Rose<Sequence<A>>> =
+  splits().map { (xs, y, zs) -> // TODO Test with discarded ones
+    y.shrinks.map { y1 -> (xs + sequenceOf(y1!!) + zs).interleave() }
+  }.asFlow().flattenConcat()
+
+internal fun <A> Sequence<Rose<A>>.interleave(): Rose<Sequence<A>> =
+  Rose(
+    this.map { it.res },
+    dropSome().asFlow().onCompletion { emitAll(shrinkOne()) }
+  )
 
-    fun <K, A> Kind<M, Tuple2<K, A>>.uniqueByKey(n: Int): Kind<M, List<Kind<M, Tuple2<K, A>>>> {
-        fun go(k: Int, map: Map<K, Kind<M, Tuple2<K, A>>>): Kind<M, List<Kind<M, Tuple2<K, A>>>> =
-            if (k > 100) discard()
-            else freeze().replicate(n).flatMap {
-                val res = (map + it.map { it.bimap({ it.a }, ::identity) }.toMap())
-                if (res.size >= n) MM().just(res.values.toList())
-                else go(k + 1, res)
-            }
-        return go(0, emptyMap())
-    }
+/**
+ * Generate a [Map] from a generator that generators key-value pairs.
+ *
+ * The range limits the size of the hashmap.
+ *
+ * This generator will shrink to an empty map and also shrink both keys and values.
+ */
+fun <R, K, A> Gen<R, Pair<K, A>>.hashMap(range: Range<Int>): Gen<R, Map<K, A>> = Gen.sized { s ->
```

Comment:
Nit: shouldn't we just use `map` here since it reflects what we find in the Std and the type we return here? `Map`. Or should we return `Gen<R, kotlin.collections.HashMap<K, A>>`
 
```suggestion
fun <R, K, A> Gen<R, Pair<K, A>>.map(range: Range<Int>): Gen<R, Map<K, A>> = Gen.sized { s ->
```